Sources/NIOCore/EventLoopFuture+AssumeIsolated.swift

```diff
@@ -0,0 +1,528 @@
+        ///
+        /// - parameters:
+        ///     - orReplace: the value of the returned `EventLoopFuture` when then resolved future's value is `Optional.some()`.
+        /// - returns: an new `EventLoopFuture` with new type parameter `NewValue` and the value passed in the `orReplace` parameter.
+        @inlinable
+        func unwrap<NewValue>(
+            orReplace replacement: NewValue
+        ) -> EventLoopFuture<NewValue>.AssumeIsolated where Value == Optional<NewValue> {
+            return self.map { (value) -> NewValue in
+                guard let value = value else {
+                    return replacement
+                }
+                return value
+            }
+        }
+
+        /// Unwrap an `EventLoopFuture` where its type parameter is an `Optional`.
+        ///
+        /// Unwraps a future returning a new `EventLoopFuture` with either: the value returned by the closure passed in
+        /// the `orElse` parameter when the future resolved with value Optional.none, or the same value otherwise. For example:
+        /// ```
+        /// var x = 2
+        /// promise.futureResult.unwrap(orElse: { x * 2 }).wait()
+        /// ```
+        ///
+        /// - parameters:
+        ///     - orElse: a closure that returns the value of the returned `EventLoopFuture` when then resolved future's value
+        ///         is `Optional.some()`.
+        /// - returns: an new `EventLoopFuture` with new type parameter `NewValue` and with the value returned by the closure
+        ///     passed in the `orElse` parameter.
+        @inlinable
+        func unwrap<NewValue>(
+            orElse callback: @escaping () -> NewValue
+        ) -> EventLoopFuture<NewValue>.AssumeIsolated where Value == Optional<NewValue> {
+            return self.map { (value) -> NewValue in
+                guard let value = value else {
+                    return callback()
+                }
+                return value
+            }
+        }
+
+        /// Returns the wrapped event loop future.
+        @inlinable
+        func nonisolated() -> EventLoopFuture<Value> {
+            return self.wrapped
+        }
+    }
+
+    /// Assumes the calling context is isolated to the future's event loop.
+    @usableFromInline
+    func assumeIsolated() -> AssumeIsolated {
+        self.eventLoop.assertInEventLoop()
+        return AssumeIsolated(wrapped: self)
+    }
+}
+
+
+extension EventLoopPromise {
+    /// A struct wrapping an ``EventLoopPromise`` that ensures all calls to any method on this struct
+    /// are coming from the event loop of the promise.
+    @usableFromInline
+    struct AssumeIsolated {
+        @usableFromInline
+        let wrapped: EventLoopPromise<Value>
+
+
+        /// Deliver a successful result to the associated `EventLoopFuture<Value>` object.
+        ///
+        /// - parameters:
+        ///     - value: The successful result of the operation.
+        @inlinable
+        func succeed(_ value: Value) {
```

Comment:
No `fail` counterpart?

Reply:
We don't need `fail` counter part since that doesn't require `Sendable` values. I really wanted to avoid replicating all methods on the new types.